src/app/qgsmaptooldigitizefeature.cpp

```diff
@@ -138,21 +138,31 @@ void QgsMapToolDigitizeFeature::cadCanvasReleaseEvent( QgsMapMouseEvent *e )
       return;
     }
 
-    QgsPointXY savePoint; //point in layer coordinates
+    QgsPoint savePoint; //point in layer coordinates
+    bool isMatchPointZ = false;
     try
     {
       QgsPoint fetchPoint;
       int res;
       res = fetchLayerPoint( e->mapPointMatch(), fetchPoint );
+      if ( QgsWkbTypes::hasZ( fetchPoint.wkbType() ) )
+        isMatchPointZ = true;
+
       if ( res == 0 )
       {
-        savePoint = QgsPointXY( fetchPoint.x(), fetchPoint.y() );
+        if ( isMatchPointZ )
+          savePoint = QgsPoint( QgsWkbTypes::PointZ, fetchPoint.x(), fetchPoint.y(), fetchPoint.z() );
+        else
+          savePoint = QgsPoint( fetchPoint.x(), fetchPoint.y() );
       }
       else
       {
-        savePoint = toLayerCoordinates( vlayer, e->mapPoint() );
+        QgsPointXY layerPoint = toLayerCoordinates( vlayer, e->mapPoint() );
```

Comment:
It might be worth to add a QgsPoint version of this method too?

Reply:
Yes, but since it's only a bugfix (and backported), I didn't want to add a new method